File: Normal/LongestSubstringWithoutRepeatingCharacters.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        length = len(s)
        if(length == 0):
            return 0
        elif(length == 1):
            return 1
        p1 = 0
        p2 = 1
        max = 1
        d = {}
        d[s[0]] = 1
        while(p2 < length):
            char = s[p2]
            while d.get(char, 0) == 1:
                d[s[p1]] -= 1
                p1 += 1
            d[char] = 1
            p2 += 1
            val = p2 - p1
            max = max if max > val else val
        return max

File: Normal/Longest Substring without Repeating Characters.rs
use std::collections::{
    BTreeMap,   // Sorted Map
    BTreeSet,   // Sorted Set
    BinaryHeap, // Max Heap Implementation
    HashMap,    // Map
    HashSet,    // Set
    LinkedList, // Linked List not to be used anytime
    VecDeque,   // Deque
};

use std::cmp;

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut longest = 0;
        let mut i = 0;
        let mut j = 0;
        let mut set = HashSet::new();
        let chars: Vec<char> = s.chars().into_iter().collect();
        while j < chars.len() {
            // println!("\n{} {}", i, j);
            // println!("1: {:?}", set);
            if i == j && i == 0 {
                set.insert(chars[i]);
                j += 1;
                longest = cmp::max(longest, j - i);
                continue;
            }
            if let Some(x) = set.get(&chars[j]) {
                while chars[i] != chars[j] {
                    // println!("----2: {:?}", set);
                    // println!("----{} {}", chars[i], chars[j]);
                    set.remove(&chars[i]);
                    i += 1
                }
                i += 1;
                j += 1;
            } else {
                set.insert(chars[j]);
                j += 1;
            }
            longest = cmp::max(longest, j - i);
        }
        longest as i32
    }
}

impl Solution {
    pub fn length_of_longest_substring(s: String) -> i32 {
        let mut longest = 0;
        let mut i = 0;
        let mut j = 0;
        let mut set = HashMap::new();
        let chars: Vec<char> = s.chars().into_iter().collect();
        while j < chars.len() {
            if let Some(x) = set.get_mut(&chars[j]) {
                i = cmp::max(i, *x);
            }
            longest = cmp::max(longest, j - i + 1);
            set.insert(chars[j], j + 1);
            j += 1;
            // println!("{} {} {:?}", i, j, set);
        }
        longest as i32
    }
}
